gobblin-data-management/src/main/java/org/apache/gobblin/data/management/source/LoopingDatasetFinderSource.java

```diff
@@ -79,59 +88,122 @@ public LoopingDatasetFinderSource(boolean drilldownIntoPartitions) {
     return Lists.newArrayList(getWorkunitStream(state).getMaterializedWorkUnitCollection());
   }
 
+  /**
+   * Collect all failed workunit in previous run into a set.
+   *
+   * Workunits that exceed max retry limit will not be added into failedPreviousWorkUnits
+   * and rollovered into next quota-available execution.
+   *
+   * @param retriedUrns Track all dataset/partition's urns that is included in the retried workunits, so that in the
+   *                    case where the number of remaining dataset/partitions are less than the maximum number of
+   *                    workunits can be processed by single execution, a retried workunit won't be processed twice
+   */
+  private List<WorkUnit> getPreviousFailedWorkUnits(List<WorkUnitState> previousWorkUnitStates, int maxRetries,
+      Set<String> retriedUrns) {
+    List<WorkUnit> failedPreviousWorkUnits = new ArrayList<>();
+    for (WorkUnitState workUnitState : previousWorkUnitStates) {
+      if ((workUnitState.getWorkingState() == WorkUnitState.WorkingState.FAILED
+          || workUnitState.getWorkingState() == WorkUnitState.WorkingState.ROLLOVER)) {
+        if (!workUnitState.contains(ConfigurationKeys.TASK_RETRIES_KEY)
+            || workUnitState.getPropAsInt(ConfigurationKeys.TASK_RETRIES_KEY) < maxRetries + 1) {
```

Comment:
Why is there a `+ 1` here? I would expect the `maxRetries` passed in to this method to be used unchanged.

Reply:
Then it should be <= here.

Reply:
No actually, my mistake. I should add a `currentRetryCount` to ensure that is retry limit is set to 0, no retry will be conducted.